### PR TITLE
feature/create-sample

### PR DESCRIPTION
**Pull request by wolfbiter**
add file drop component to samples list. add createSample reducers on file drop

Ok, two architecture questions on this one.

1) createSample must be async, as the id is calculated from a checksum on the file. so i thought we could go one of two ways - either load that sample straight into the store, or copy the file first then just call `loadSample` action. i went with the former but was unsure - wdyt? easy enough to change, so want to pick the right thing

2) how should we handle the case of de-duplication? would it make sense, in `service.createSample`, to return the existing sample if there is one already? or what is the correct approach there? the sequencing here is important, because i think the desired UX is that when you drop a file onto the page, you want it to analyze in the background. but if that file is already in the system, it should use the existing copy and NOT override the metadata by triggering a new analysis

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -38,6 +39,7 @@
     "pify": "^2.3.0",
     "react": "^15.4.0",
     "react-dom": "^15.4.0",
+    "react-file-drop": "^0.1.8",
```
Comment (wolfbiter) [inline package.json:42]: what do i / dont i add to package? are these right?

File: samples/getters.js [hunk 2]
```diff
@@ -3,21 +3,29 @@ const { values } = require('lodash')
 
 const getSamples = (state) => state.samples.records
 const getSamplesIsLoading = (state) => state.samples.isLoading
+const getSamplesIsCreating = (state) => state.samples.isCreating
+const getSamplesError = (state) => state.samples.error
 
 const getSampleList = Getter(
   getSamples,
+
+  // TODO: load and connect metas, too. display list of titles, not IDs
```
Comment (wolfbiter) [inline samples/getters.js:12]: in the future would be nice. not gonna do now

File: samples/containers/sample-list.js [hunk 3]
```diff
@@ -1,17 +1,36 @@
...
+    const { sampleList, isLoading, isCreating, error } = this.props
 
     return <div>
+      <FileDrop
+        frame={document}
+        onFrameDrop={this.onFilesDrop.bind(this)} />
       <header>
-        samples are {isLoading ? 'loading' : 'here'}
+        <div>samples are {isLoading ? 'loading' : 'here'}</div>
+        {isCreating && <div>'creating samples…'</div>}
```
Comment (wolfbiter) [inline samples/containers/sample-list.js:32]: seems like at this point i might need to do the fancy `creatingSamples: [{ id }, ... ]`. any easy way of implementing that? havent thought it through

File: samples/reducer.js [hunk 4]
```diff
@@ -54,9 +68,47 @@ function createReducer (config) {
...
+      }
+
+      return loop({
+        ...state,
+        records: {
+          ...state.records,
+          [action.payload.id]: action.payload
+        }
+      }, Effects.batch([
+        // TODO: here is a case where ordering might be nice.
```
Comment (wolfbiter) [inline samples/reducer.js:97]: thoughts? should i always be creating my effects in a way that doesnt need a specific order?

File: samples/containers/sample-list.js (hunk 3 above)
Comment (wolfbiter) [reply samples/containers/sample-list.js:32]: i will def need it in the case of `analyzingSamples: [sample1, sample2, ...]`

File: samples/reducer.js [hunk 5]
```diff
@@ -54,9 +68,47 @@ function createReducer (config) {
...
+    }),
+    [createSample]: (state, action) => {
+      const file = action.payload
+      assert(file && file.path, 'Cannot createSample without file && file.path')
+
+      return loop({
+        ...state, isCreating: true
+      }, Effects.batch([
+        Effects.promise(runCreateSample, file),
+        Effects.constant(createSampleEnd())
```
Comment (wolfbiter) [inline samples/reducer.js:80]: note i tried the alternative approach in [the next pr](https://github.com/linxmix/linx-electron/pull/11/files#diff-def1f84f5412306bde74c0bb223f28d7R49), copying the file first then just calling `readSample`

File: package.json [hunk 6]
```diff
@@ -29,6 +29,7 @@
     "babel-preset-react": "^6.16.0",
     "babel-register": "^6.18.0",
     "core-js": "^2.4.1",
+    "crypto": "0.0.3",
```
Comment (ahdinosaur) [inline package.json:32]: no need to add this, it's built-in to node.

File: package.json (hunk 1 above)
Comment (ahdinosaur) [reply package.json:42]: yes add this.

File: samples/containers/sample-list.js (hunk 3 above)
Comment (ahdinosaur) [reply samples/containers/sample-list.js:32]: store whether or not the sample is being created or analyzed somewhere in the store, then in the getters attach this information to the samples and then `analyzingSamples: [sample1, sample2, ...]` is just a filtered list where `analyzing: true`.

File: samples/reducer.js (hunk 4 above)
Comment (ahdinosaur) [reply samples/reducer.js:97]: nah, in this case you want the effects to be called in order, like a `.then`.

hmm, nest the calls to batch? https://github.com/redux-loop/redux-loop/blob/v2.2.2/docs/ApiDocs.md#effectsbatcheffects

/cc @iainkirkpatrick @gregorykan what's the best way to run many `redux-loop` effects in order?

**Comment by ahdinosaur** [comment]
> 1) createSample must be async, as the id is calculated from a checksum on the file. so i thought we could go one of two ways - either load that sample straight into the store, or copy the file first then just call `loadSample` action. i went with the former but was unsure - wdyt? easy enough to change, so want to pick the right thing

*shrug*. i'd probably do the latter but i don't think there's a "right" way.

> 2) how should we handle the case of de-duplication? would it make sense, in `service.createSample`, to return the existing sample if there is one already? or what is the correct approach there? the sequencing here is important, because i think the desired UX is that when you drop a file onto the page, you want it to analyze in the background. but if that file is already in the system, it should use the existing copy and NOT override the metadata by triggering a new analysis

if the checksum is the same, yeah return the sample and skip the rest.

**Review by ahdinosaur — APPROVED** [review]

File: samples/containers/sample-list.js (hunk 3 above)
Comment (wolfbiter) [reply samples/containers/sample-list.js:32]: tracking as future work

**Pull request merged**